packages/snap/src/snap.ts

```diff
@@ -9,13 +9,39 @@ import {getBlock} from "./rpc/substrate/getBlock";
 import {removeAsset, updateAsset} from "./asset";
 import {getApi} from "./polkadot/api";
 import {configure} from "./rpc/configure";
+import {polkadotEventEmitter} from "./polkadot/events";
+import {registerOnBalanceChange} from "./polkadot/events/balance";
+import {EventCallback, PolkadotApi, PolkadotEvent} from "@nodefactory/metamask-polkadot-types";
 
 declare let wallet: Wallet;
 
 const apiDependentMethods = ["getBlock", "getBalance", "getChainHead", "addKusamaAsset"];
 
+wallet.registerApiRequestHandler(async function (origin: string): Promise<PolkadotApi> {
```

Comment:
@mpetrunic It is not possible just to move this function call inside `on` because then we will have multiple calls of this function that will result in multiple calls of **emit** when the balance changes.

I think there are two possible solutions:
1) `Call this register here`
:+1: No additional code
:-1: If there are no listeners, **emit** will be invoked on _balance change_ (on empty array so nothing will happen) 

2) `Add additional logic to track the first call and then subscribe on balance change, additionally, it should unregister when all listeners are removed`
:+1: No "empty" **emit** calls
:-1: Adding additional logic

Reply:
It's not just emit calls, you are wasting cpu cycles of metamask by keeping connection open and listening on balance. So I think we should merge this once you open issues:
- subscribe on first call and unregister when all listeners are removed
- don't create polkadot asset automatically (maybe that isn't in code anymore) and subscribe on balance change after it's added

Reply:
Issues are opened #57 #56, just approve so I close PR.